src/module.ts

```diff
@@ -1102,6 +1104,14 @@ export class Module {
   ) {
     assert(sizeType == TypeRef.I32 || sizeType == TypeRef.I64);
     this.lit = binaryen._malloc(binaryen._BinaryenSizeofLiteral());
+
+    // cache heap types
+    TypeRef.Funcref = binaryen._BinaryenTypeFuncref();
+    TypeRef.Externref = binaryen._BinaryenTypeExternref();
+    TypeRef.Anyref = binaryen._BinaryenTypeAnyref();
+    TypeRef.Eqref = binaryen._BinaryenTypeEqref();
+    TypeRef.I31ref = binaryen._BinaryenTypeI31ref();
+    TypeRef.Dataref = binaryen._BinaryenTypeDataref();
```

Comment:
I guess currently this works because there's always a module construction before other things. That's not necessarily guaranteed, though. Ideas how to cover other cases as well?

Reply:
I'm not sure there's anything else we can do in the other way. A more radical option is to refactor from global namespace to instance-based approach where mode will aggregate TypeRefs and expose them via getters for example.

Reply:
I *think* we'll have to move types into `Program` context one day, so if we can't do anything else now, that would probably resolve it as well. Ideally, though, if the above issue could be resolved, we'd gain a bit more time.